File: java/117_PopulatingNextRightPointersInEachNodeII.java
class Solution {
    Node prev, leftmost;
    public Node connect(Node root) {
        if (root == null) return root;
        leftmost = root;
        Node curr;
        while (leftmost != null) {
            curr = leftmost;
            prev = leftmost = null;
            while (curr != null) {
                processChild(curr.left);
                processChild(curr.right);
                curr = curr.next;
            }
        }
        return root;
    }
    
    public void processChild(Node child) {
        if (child != null) {
            if (prev != null) {
                prev.next = child;
            } else {
                leftmost = child;
            }
            prev = child;
        }
    }
}

class Solution {
    public Node connect(Node root) {
        if (root == null)
            return null;
        
        Deque<Node> q = new ArrayDeque<Node>(){{
                addFirst(root);
        }};
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                Node node = q.pollFirst();
                if (node.left != null)
                    q.addLast(node.left);
                if (node.right != null)
                    q.addLast(node.right);
                
                node.next = i == size - 1 ? null : q.peekFirst();
            }
        }
        return root;
    }
}

File: swift/117_PopulatingNextRightPointersInEachNodeII.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var left: Node?
 *     public var right: Node?
 *	   public var next: Node?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func connect(_ root: Node?) -> Node? {
        func processChild(_ currentChildNode: Node?) {
            var child = currentChildNode
            if child != nil {
                if prev != nil {
                    prev?.next = child
                } else {
                    leftmost = child
                }
                prev = child
            }
        }
        
        var prev: Node? = nil, leftmost: Node? = root
        while leftmost != nil {
            var curr: Node? = leftmost
            leftmost = nil
            prev = nil
            while curr != nil {
                processChild(curr?.left)
                processChild(curr?.right)
                curr = curr?.next
            }
        }
        return root
    }
}
